python/tritonserver/_c/tritonserver_pybind.cc

```diff
@@ -1175,21 +1162,109 @@ class PyInferenceRequest
     ThrowIfError(TRITONSERVER_InferenceRequestSetDoubleParameter(
         triton_object_, key.c_str(), value));
   }
+
   void Cancel()
   {
     ThrowIfError(TRITONSERVER_InferenceRequestCancel(triton_object_));
   }
 
+  // Response management
+  void GetNextResponse(const py::object& py_future)
+  {
+    // Called exclusively from Python, so GIL is always held.
+    std::lock_guard lock(response_mu_);
 
- public:
-  std::unique_ptr<CallbackResource> request_callback_resource_{nullptr};
+    // There are two cases:
+    // 1. The backend is faster than the frontend:
+    //      There will be response(s) queued at 'responses_' deque, so pop the
+    //      response from the deque and set it to the 'py_future' object.
+    // 2. The backend is slower than the frontend:
+    //      The 'responses_' deque will be empty, so track the 'py_future' in
+    //      this request object until the 'AddNextResponse()' is called.
+    if (responses_.empty()) {
+      if (response_future_.ptr() != nullptr) {
+        throw AlreadyExistsError("cannot call GetNextResponse concurrently");
+      }
+      response_future_ = py_future;
```

Comment:
question - does this increment the py_future ref count?

Reply:
yes, the ref count of `response_future_` is decremented, if it is managing an object, and the `py_future` ref count is incremented and the additional count is managed by the `response_future_`.
https://github.com/pybind/pybind11/blob/d2e7e8c68711d1ebfb02e2f20bd1cb3bfc5647c0/include/pybind11/pytypes.h#L405-L416